clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java

```diff
@@ -155,7 +155,7 @@ public boolean maybeExpire(int requestTimeoutMs, long retryBackoffMs, long now,
 
         if (!this.inRetry() && isFull && requestTimeoutMs < (now - this.lastAppendTime))
             expiryErrorMessage = (now - this.lastAppendTime) + " ms has passed since last append";
-        else if (!this.inRetry() && requestTimeoutMs < (now - (this.createdMs + lingerMs)))
+        else if (!this.inRetry() && requestTimeoutMs < (now - Math.max(this.createdMs + lingerMs, lingerMs)))
             expiryErrorMessage = (now - (this.createdMs + lingerMs)) + " ms has passed since batch creation plus linger time";
         else if (this.inRetry() && requestTimeoutMs < (now - (this.lastAttemptMs + retryBackoffMs)))
```

Comment:
We should probably follow the same pattern here and in the respective error message.